README.md
# SRE Challenge
## Simple Checkout Basket in golang

A simple checkout HTTP API that contains below operations:
- Create a new checkout basket
- Add a product to a basket
- Get the total amount in a basket
- Remove the basket

The store only sells following 3 products:

```bash
Code         | Name              |  Price
-----------------------------------------------
PEN          | Lana Pen          |   5.00€
TSHIRT       | Lana T-Shirt      |  20.00€
MUG          | Lana Coffee Mug   |   7.50€
```

Various departments have insisted on the following discounts:

- The sales department thinks a buy 2 get 1 free promotion will work best (for each two of the same product, get the second free), and would like this to only apply to PEN items.
- The CFO insists that the best way to increase sales is with discounts on bulk purchases (buying x or more of a product, the price of that product is reduced), and requests that if you buy 3 or more TSHIRT items, the price per unit should be reduced by 25%.

Examples:

```bash
Items: PEN, TSHIRT, MUG
Total: 32.50€

Items: PEN, TSHIRT, PEN
Total: 25.00€

Items: TSHIRT, TSHIRT, TSHIRT, PEN, TSHIRT
Total: 65.00€

Items: PEN, TSHIRT, PEN, PEN, MUG, TSHIRT, TSHIRT
Total: 62.50€
```

## Requirements
- Golang v1.10.0+
- Docker

## Installation

### Manual Installation (local)

#### (Optional environment variable)
The api using a simple json db (scribble golang module). And it has an optional environment variable "DB_NAME".
Default value for DB_NAME=basket_db, if you want to change it into something else you can set the variable

```bash
$ export DB_NAME=<your-db-name>
```

```bash
$ git clone https://github.com/stvnorg/golang-checkout-basket
$ cd golang-checkout-basket
$ export DB_NAME=<LOCAL_JSON_DB> (optional)
$ go run api.go basket.go discounts.go
```
If you run it the first time, it will take some times for the go dependency manager to install the libraries needed. So, be patient! wait a little bit :)
The API has Prometheus enabled. It will listen/expose 2 TCP ports, 8080 (the API itself) and 9900 (prometheus exporter)

### Docker run
The docker image for the api is hosted in docker hub https://hub.docker.com/r/s7even/golang-checkout-basket
to run it simply by executing below command (without prometheus exposed):

```bash
$ docker run -p 8080:8080 -d s7even/golang-checkout-basket:0.2.2
```

and if you want to expose both with prometheus port, just run below command:

```bash
$ docker run -p 8080:8080 -p 9900:9900 -d s7even/golang-checkout-basket:0.2.2
```
Browse the http://localhost:8080/basket/total, to see if the API is running well

## Unit Test

To run the unit test (verbose):
```bash
$ go test -v
```

Unit test with coverage:
```bash
$ go test -cover
```

## Usage

Before you able to add products, you need to create a basket first

#### Create basket
```bash
$ curl -X POST http://<API_URL>/basket/create
```
Example:
```bash
$ curl -X POST http://localhost:8080/basket/create
```

#### Add Product to basket
```bash
$ curl -X POST http://<API_URL>/basket/add_product/<pen|mug|tshirt>
```
Example:
```bash
$ curl -X POST http://localhost:8080/basket/add_product/pen
```

#### Get the total amount in basket
```bash
$ curl -X GET http://<API_URL>/basket/total
```
Example:
```bash
$ curl -X GET http://localhost:8080/basket/total
```

#### Remove the basket
```bash
$ curl -X DELETE http://<API_URL>/basket/delete
```
Example:
```bash
$ curl -X DELETE http://localhost:8080/basket/delete
```

## CI Pipeline process
There are 2 (two) githubActions CI pipeline:

`.github/workflows/go.yml`
- Developers create a Pull Request and/or Merged the PR
- The CI will run the build and unit test

`.github/workflows/release.yml`
- Developers push new tags with semVer versioning X.Y.Z (ex: 1.0.0)
- The CI pipeline will login to docker hub
- Build the image with tagging `s7even/golang-checkout-basket:<X.Y.Z>`
- Push the image to docker hub

Note: Make sure to update the CHANGES file for the release notes

## (Optional) Prometheus Monitoring

The prometheus exporter is running on port 9900
If you choose to enabling it in docker run, make sure you add the correct targets in your `prometheus.yml`

Example:
```bash
global:
  scrape_interval:     15s
  evaluation_interval: 15s

scrape_configs:
  - job_name: prometheus
    static_configs:
      - targets: ['192.168.0.10:9900']
```


CHANGES
0.2.2 (2021-02-07)
------------------
* Standarize the json response data
* Update the README.md

0.2.1 (2021-02-07)
------------------
* Bug fix for the basket.go file

0.2.0 (2021-02-06)
------------------
* Finalized the README.md

0.1.9 (2021-02-06)
------------------
* Simplified the discount calculation formula
* Add Image digest in the ci pipeline (for security reason)
* Add Prometheus monitoring

0.1.8 (2021-02-06)
------------------
* Add CI Release pipeline (build and push docker image)

0.1.7 (2021-02-05)
------------------
* Refactor the code
* Add Dockerfile

0.1.6 (2021-02-04)
------------------
* Add unittest for the api routes

0.1.5 (2021-02-04)
------------------
* Bug fix for the api router

0.1.4 (2021-02-04)
------------------
* Add unit test
* Bug fix for the discounts calculation

0.1.3 (2021-02-03)
------------------
* Add the route for adding_products

0.1.2 (2021-02-03)
------------------
* Add routes for Create and Delete basket

0.1.1 (2021-02-03)
------------------
* Implementing Gin rest-api module
* Add route to get basket total amount http://api_url/basket/total

0.1.0 (2021-02-02)
------------------
* Project Initialization, golang Rest-API with simple json db (scribble)


api_test.go
package main

import (
  "github.com/gin-gonic/gin"
  "net/http/httptest"
  "testing"
)

func Test_GET_CustomHTTP404(t *testing.T) {
  w := httptest.NewRecorder()
  c, _ := gin.CreateTestContext(w)

  GET_CustomHTTP404(c)
  if w.Code != 404 {
    t.Fail()
  }
}

func Test_POST_CreateBasket(t *testing.T) {
  w := httptest.NewRecorder()
  c, _ := gin.CreateTestContext(w)

  POST_CreateBasket(c)
  if w.Code != 201 {
    t.Fail()
  }
}

func Test_POST_AddProductToBasket(t *testing.T) {
  w := httptest.NewRecorder()
  c, _ := gin.CreateTestContext(w)
  c.Params = []gin.Param{gin.Param{Key: "code", Value: "v"}}

  POST_AddProductToBasket(c)
  if w.Code != 200 {
    t.Fail()
  }
}

func Test_GET_TotalAmountInBasket(t *testing.T) {
  w := httptest.NewRecorder()
  c, _ := gin.CreateTestContext(w)

  GET_TotalAmountInBasket(c)
  if w.Code != 200 {
    t.Fail()
  }
}

func Test_DELETE_DeleteBasket(t *testing.T) {
  w := httptest.NewRecorder()
  c, _ := gin.CreateTestContext(w)

  DELETE_DeleteBasket(c)
  if w.Code != 200 {
    t.Fail()
  }
}


basket_test.go
package main

import "testing"

import (
	"github.com/nanobox-io/golang-scribble"
)

var test_db, _ = scribble.New("./" + GetEnv("DB_NAME", "test_db"), nil)

func Test_CreateBasket(t *testing.T) {
    CreateBasket()
}

func Test_BasketExist(t *testing.T) {
  if !IsBasketExist() {
    t.Fail()
  }
}

func Test_AddProductToBasket(t *testing.T) {
	AddProductToBasket("PEN")
}

func Test_GetBasketTotalAmount(t *testing.T) {
	test_basket := GetBasketTotalAmount()
	if test_basket.Total != 5.0 {
		t.Fail()
	}
}

func Test_DeleteBasket(t *testing.T) {
	DeleteBasket()
}


discounts_test.go
package main

import "testing"

func Test_DiscountPrice(t *testing.T) {
  if DiscountPrice([]string{"PEN", "TSHIRT", "MUG"}) != 32.50 {
    t.Fail()
  }

  if DiscountPrice([]string{"PEN", "TSHIRT", "PEN"}) != 25.00 {
    t.Fail()
  }

  if DiscountPrice([]string{"TSHIRT", "TSHIRT", "TSHIRT", "PEN", "TSHIRT"}) != 65.00 {
    t.Fail()
  }

  if DiscountPrice([]string{"PEN", "TSHIRT", "PEN", "PEN", "MUG", "TSHIRT", "TSHIRT"}) != 62.50 {
    t.Fail()
  }
}


main.go
package main

import (
    "fmt"
    "github.com/gin-gonic/gin"
    "strings"
)

var message string

func main() {
  r := gin.Default()
  r.NoRoute(func(c *gin.Context) {
    c.JSON(404, gin.H{"message": "Page not found"})
  })

  r.POST("/basket/create", func(c *gin.Context) {
    CreateBasket()

    if IsBasketExist() {
      message = "Basket has been created"
    } else {
      message = "Error creating basket"
    }

    c.JSON(201, gin.H{
      "message": message,
    })
  })


  r.POST("/basket/add_product", func(c *gin.Context) {
    code := strings.ToUpper(c.Query("code"))

    if code == "MUG" || code == "PEN" || code == "TSHIRT" {
      AddProductToBasket(code)
      message = "Product successfully added to the basket"
    } else {
      message = "Product code doesn't exist"
    }

    c.JSON(200, gin.H{
      "message": message,
    })
  })

  r.GET("/basket/total", func(c *gin.Context) {
    if IsBasketExist() {
      mybasket := GetBasketTotalAmount()
      c.JSON(200, gin.H{
        "products": mybasket.Products,
        "total_amount": fmt.Sprintf("%.2f%s", mybasket.Total, "€"),
      })
    } else {
      c.JSON(200, gin.H{
        "message": "Basket doesn't exist",
      })
    }
  })

  r.DELETE("/basket/delete", func(c *gin.Context) {
    if DeleteBasket() {
      message = "Basket has been deleted"
    } else {
      message = "Error deleting basket"
    }

    c.JSON(200, gin.H{
      "message": message,
    })
  })

  r.Run()
}


basket.go
package main

import (
	"github.com/nanobox-io/golang-scribble"
	"os"
)

type Basket struct {
	Name string
	Products []string
	Total float64
}

func GetEnv(key, fallback string) string {
	value, exists := os.LookupEnv(key)
	if !exists {
		value = fallback
	}
	return value
}

// retrieve the environment variable DB_NAME
var db_name = GetEnv("DB_NAME", "basket_db")

// Initialize the scribble simple json db with name db_name
var db, _ = scribble.New("./" + db_name, nil)

var mybasket = Basket{}

// Check if basket exist before adding products
func IsBasketExist() bool {
	db.Read(db_name, "basket", &mybasket)
	return mybasket.Name != ""
}

// Create the basket
func CreateBasket() {
	mybasket.Name = "mybasket"
	mybasket.Products = []string{}
	db.Write(db_name, "basket", mybasket)
}

// Get the basket
func GetBasketTotalAmount() Basket {
	db.Read(db_name, "basket", &mybasket)
	return mybasket
}

// Add product to basket
func AddProductToBasket(product_name string) {
	db.Read(db_name, "basket", &mybasket)
	mybasket.Products = append(mybasket.Products, product_name)
	mybasket.Total = DiscountPrice(mybasket.Products)
	db.Write(db_name, "basket", mybasket)
}

// Delete the basket
func DeleteBasket() bool {
	db.Delete(db_name, "")
	mybasket.Name = ""
	mybasket.Products = []string{}
	mybasket.Total = 0.0
	return !IsBasketExist()
}


api.go
package main

import (
    "fmt"
    "github.com/gin-gonic/gin"
    "github.com/banzaicloud/go-gin-prometheus"
    "strings"
)

var message string

// Route for custom http error 404
func GET_CustomHTTP404(c *gin.Context) {
  c.JSON(404, gin.H{"message": "Page not found"})
}

// Route for creating a basket
func POST_CreateBasket(c *gin.Context) {
  CreateBasket()

  if IsBasketExist() {
    message = "Basket has been created"
  } else {
    message = "Error creating basket"
  }

  c.JSON(201, gin.H{
    "message": message,
  })
}

// Route for adding a product to basket
func POST_AddProductToBasket(c *gin.Context) {
  if IsBasketExist() {
    code := strings.ToUpper(c.Param("code"))
    if code == "MUG" || code == "PEN" || code == "TSHIRT" {
      AddProductToBasket(code)
      message = "Product successfully added to the basket"
    } else {
      message = "Product code doesn't exist"
    }
  } else {
    message = "Basket doesn't exist"
  }

  c.JSON(200, gin.H{
    "message": message,
  })
}

// Route for getting the total amount in basket
func GET_TotalAmountInBasket(c *gin.Context) {
  if IsBasketExist() {
    mybasket := GetBasketTotalAmount()
    c.JSON(200, gin.H{
      "products": mybasket.Products,
      "total_amount": fmt.Sprintf("%.2f%s", mybasket.Total, "€"),
    })
  } else {
    c.JSON(200, gin.H{
      "message": "Basket doesn't exist",
    })
  }
}

// Route to remove basket
func DELETE_DeleteBasket(c *gin.Context) {
  if DeleteBasket() {
    message = "Basket has been deleted"
  } else {
    message = "Error deleting basket"
  }

  c.JSON(200, gin.H{
    "message": message,
  })
}

func main() {
  r := gin.Default()

  p := ginprometheus.NewPrometheus("gin", []string{})
  p.SetListenAddress(":9900")
  p.Use(r, "/metrics")

  r.NoRoute(GET_CustomHTTP404)
  r.POST("/basket/create", POST_CreateBasket)
  r.POST("/basket/add_product/:code", POST_AddProductToBasket)
  r.GET("/basket/total", GET_TotalAmountInBasket)
  r.DELETE("/basket/delete", DELETE_DeleteBasket)

  r.Run()
}
